docs/articles/rag-application-communication-system.md

```diff
@@ -0,0 +1,164 @@
+
+Fine-tuning is better conceived as a method for enhancing communication, by tweaking the word probabilities of the base model in a specific direction: communication between the model and the user, communication between the resources and the output and, as agent-based approaches are emerging, between different models.
+
+### 4.1 Fine-tuning for/through RAG
+Fine-tuning has to come later in the cycle of development. In my experience, good instruction datasets need to be designed not only with RAG in mind but through RAG. My standard protocol of fine-tuning for RAG goes this way:
+* Generate queries based on indexed texts. This is a really standard and basic approach of synthetic data universally used for fine-tuning: given a text, invent a question it would answer. In this specific context, it is recommended to prompt the model to create “vague” enough questions that could yield more valid answers than the seeding text. Also to go one step further, you’d want the queries to match the ones sent by your user. I recommend including a random selection 
+* Retrieve a selection of documents. Here it may be actually good to sometimes voluntarily degrade the performance of retrieval (for instance using only BM25), as you want the LLM to be more resilient to hard negatives.
+* Generate the actual synthesis based on the question and the retrieved documents. This is the most creative part by far, the one that actually controls the final communication output, the end behavior of the model. You don’t have to get it right in one pass. Since LORA fine-tuning is cheap anyway, I start with a simple RAG prompt on an existing model and then continuously reshape and rework this output.
+
+Fine-tuning does not require a lot of examples. Most of my projects are in the 1,000-4,000 instructions range, sufficiently large to warrant text generation, sufficiently small to still be largely an opinionated product. There is no way around it: while LLMs are all about text automation, a good fine-tuning dataset will require a significant amount of careful manual tweaking.
+
+The instruction dataset is in many ways a miniature version of the RAG communication system. You need to approximate as much as possible the future queries your users will send (so preferably based on the past ones), the existing and future documents you will use for grounding and the expected output. In short, you are designing a translation process and, through the instructions, providing as many use cases as possible where this translation went right.
+
+Otherwise, I would not spend so much time optimizing the training design beyond a few hyperparameters (learning rate, batch size, etc.). Data work and base model improvement had always the most impact. Typically, I stopped looking into preference fine-tuning (like DPO) as the time spent in additional data preparation was not worth the very few improvement points.
```

Comment:
otherwise = besides doing what?

Reply:
Yes can be dropped. Or "while the preparation of the instruction dataset should be your main focus, I…"